### PR TITLE
Per track meter internals

### PR DESCRIPTION
**Pull request by Paul-Licameli**
Resolves: *(direct link to the issue)*

Depends on
- #6442

Some of this originated in #5779 

Put into AudioIO.cpp all the support needed for per-track meters displaying correct peak and RMS data
(post all adjustments for the effect stack, pan, gain, envelope).

But also other cleanups of AudioIO.  Move the experimental automated input level adjustment out, so the Meter
interface needs fewer virtual functions.

Also clean up the publication of events from AudioIO and correct some non-thread-safe message callbacks.

To do:  MixerBoard should use this.  It doesn't show correct data now in all cases.  See #5117 

*(short description of the changes and the motivation to make the changes)*

- [x] I signed [CLA](https://www.audacityteam.org/cla/)
- [x] The title of the pull request describes an issue it addresses
- [x] If changes are extensive, then there is a sequence of easily reviewable commits
- [x] Each commit's message describes its purpose and effects
- [x] There are no behavior changes unnecessary for the stated purpose of the PR

Recommended:
- [x] Each commit compiles and runs on my machine without known undesirable changes of behavior

### CODE REVIEW

File: src/widgets/MeterPanel.cpp [hunk 1]
```diff
@@ -1126,8 +1126,8 @@ void MeterPanel::OnMeterUpdate(wxTimerEvent & WXUNUSED(event))
 
    if (numChanges > 0) {
       #ifdef EXPERIMENTAL_AUTOMATED_INPUT_LEVEL_ADJUSTMENT
-         if (gAudioIO->AILAIsActive() && mIsInput && !discarded) {
-            gAudioIO->AILAProcess(maxPeak);
+         if (AudioIO::Get()->AILAIsActive() && mIsInput && !discarded) {
```
Comment (crsib) [inline src/widgets/MeterPanel.cpp]: It seems that the only place `EXPERIMENTAL_AUTOMATED_INPUT_LEVEL_ADJUSTMENT` should be used in such approach is `AILAIsActive`.

File: src/widgets/MeterPanel.cpp [hunk 2]
```diff
@@ -1126,8 +1126,8 @@ void MeterPanel::OnMeterUpdate(wxTimerEvent & WXUNUSED(event))
...
+            AudioIO::Get()->AILAProcess(maxPeak);
             putchar('\n');
```
Comment (crsib) [inline src/widgets/MeterPanel.cpp]: But this code is disturbing. Why is it here? 

This transformation is not quite conservative anyway, so probably this line should be removed?

It seems that AILA logs extensively, probably way to extensively. Is it ready for use at all?

File: src/widgets/MeterPanel.cpp [hunk 3]
```diff
@@ -830,23 +823,16 @@ void MeterPanel::SetStyle(Style newStyle)
...
-      int inputSource;
-
+   if (mSlider) {
       Refresh();
-
       auto gAudioIO = AudioIO::Get();
-      gAudioIO->GetMixer(&inputSource, &inputVolume, &outputVolume);
-
+      auto settings = gAudioIO->GetMixer();
+      auto &[inputSource, inputVolume, outputVolume] = settings;
```
Comment (crsib) [inline src/widgets/MeterPanel.cpp:715]: I think the code would be much more readable without this line.

File: libraries/lib-audio-io/AudioIO.cpp [hunk 4]
```diff
@@ -2376,7 +2377,9 @@ void AudioIO::AILAProcess(double maxPeak) {
 
          putchar('\n');
          mAILAMax = pInputMeter ? ToLinearIfDB(mAILAMax, pInputMeter->GetDBRange()) : 0.0;
-         double iv = (double) Px_GetInputVolume(mPortMixer);
+         auto settings = audioIO.GetMixer();
+         auto &[_, inputVolume, __] = settings;
```
Comment (crsib) [inline libraries/lib-audio-io/AudioIO.cpp]: Again, why?

File: libraries/lib-audio-io/AudioIO.cpp [hunk 5]
```diff
@@ -2401,14 +2404,16 @@ void AudioIO::AILAProcess(double maxPeak) {
...
                auto msg = XO(
-"Automated Recording Level Adjustment decreased the volume to %f.").Format( vol );
+"Automated Recording Level Adjustment decreased the volume to %f.")
+                  .Format(inputVolume);
                ProjectStatus::Get( *proj ).Set(msg);
                changetype = 1;
-               wxPrintf("\tnew vol:%f\n", vol);
-               float check = Px_GetInputVolume(mPortMixer);
+               wxPrintf("\tnew vol:%f\n", inputVolume);
+               auto [_, check, __] = audioIO.GetMixer();
```
Comment (crsib) [inline libraries/lib-audio-io/AudioIO.cpp]: This one is the weirdest case of structured binding use so far.

File: libraries/lib-audio-io/AudioIO.cpp [hunk 6]
```diff
@@ -2348,14 +2348,18 @@ bool AudioIO::AILAIsActive() {
...
-   wxPrintf("START TIME %f\n\n", mAILAAbsolutStartTime);
+   mAILAAbsoluteStartTime = AudioIO::Get()->GetClockTime();
+   wxPrintf("START TIME %f\n\n", mAILAAbsoluteStartTime);
 }
 
 double AudioIO::AILAGetLastDecisionTime() {
    return mAILAAnalysisEndTime;
 }
 
+double AudioIO::GetClockTime() const {
```
Comment (crsib) [inline libraries/lib-audio-io/AudioIO.cpp:2287]: `GetStreamTime` returns stream time specific to Audacity, `GetClockTime` - actual stream time. Quite confusing, and I don't understand why is it called "clock" time.

File: libraries/lib-audio-io/AudioIO.cpp [hunk 7]
```diff
@@ -224,8 +224,12 @@ bool AudioIO::ValidateDeviceNames(const wxString &play, const wxString &rec)
    return pInfo != nullptr && rInfo != nullptr && pInfo->hostApi == rInfo->hostApi;
 }
 
+static constexpr auto kTimerInterval = std::chrono::milliseconds{ 50 };
+
 AudioIO::AudioIO()
 {
+   wxTimer::Start(kTimerInterval.count());
```
Comment (crsib) [inline libraries/lib-audio-io/AudioIO.cpp:218]: There is no safety around what kind of duration the kTimerInterval will be. Explicit `duration_cast<milliseconds>` would help.

**Comment by Paul-Licameli** [comment]
I should have put the big red Experimental label on this bit of holiday hackathon.  I didn't ask review yet, but thank you.

**Comment by Paul-Licameli** [comment]
@saintmatthieu @vsverchinsky 

Rebased onto latest master.

There is much cleanup of AudioIO.cpp that I prepared.  It's still not all of the cleanup.

Correct inter-thread communication of the stream of most recently played samples is necessary for a per-track meter display that is not meaningless garbage.

**Pull request closed without merging**